### PR TITLE
Validate forms, sections and cdes when exporting

### PR DESCRIPTION
**Pull request by alintreznai**
Validate forms, sections and CDEs when exporting registries.

### CODE REVIEW

File: rdrf/rdrf/services/io/defs/exporter.py [hunk 1]
```diff
@@ -145,6 +161,8 @@ def _create_section_map(self, section_code, optional=False):
             if optional:
                 return {}
             raise
+        if not optional:
```
Comment (sztamas) [inline rdrf/rdrf/services/io/defs/exporter.py]: Why is the if necessary?
If the section didn't exist but was optional we already returned `{}` and won't get to this point.

File: rdrf/rdrf/services/io/defs/exporter.py [hunk 2]
```diff
@@ -112,6 +121,13 @@ def _get_cdes(self, export_type):
         generic_cdes = self._get_generic_cdes()
         return self._sort_codes(cdes.union(generic_cdes))
 
+    def _check_model_validity(self, model):
+        try:
+            model.clean()
+        except ValidationError as verr:
+            raise ExportException(verr.message)
```
Comment (sztamas) [inline rdrf/rdrf/services/io/defs/exporter.py]: Maybe worth adding `from verr` to keep the `ValidationError`? Might be useful.
Probably should be `str(verr)` as I don't think `verr.message` is always a property of the validation error.

Also, the message should contain the model to give us better error messages.
But even if you do that the Export management command has to be changed, to display more helpful error messages. Currently, if the export fails you don't know why it did so.

**Review by sztamas — COMMENTED** [review]
The approach might work ok and it is helpful as it is if we improved the error messages, so we see why the validation failed.

However, it seems a bit hackish. What is the purpose of the new Export Format VALIDATION?
Is it just so we can run validation of a registry by using the Export Format?
If so, I'm not sure I like it. We can just use format None or YAML and ignore the exported data.
Sort of a `--simulate` or `--dry-run`.

As discussed, I would be more inclined to have another class responsible for validating a registry.
I do see that this approach is way easier to implement though and it is better than the current export, so I think it is fine to keep it as long as we improve the error messages and maybe drop the new Export Format.

File: rdrf/rdrf/services/io/defs/exporter.py (hunk 2 above)
Comment (alintreznai) [reply rdrf/rdrf/services/io/defs/exporter.py]: Updated to include the model in the exception message and wrap the original exception.

File: rdrf/rdrf/services/io/defs/exporter.py (hunk 1 above)
Comment (alintreznai) [reply rdrf/rdrf/services/io/defs/exporter.py]: Fixed.

**Comment by alintreznai** [comment]
Updated the code so that validation is separated from the export and it's configurable via an export parameter.

File: rdrf/rdrf/services/io/defs/exporter.py [hunk 3]
```diff
@@ -54,7 +59,39 @@ class Exporter(object):
...
+        except Section.DoesNotExist:
+            raise
+        self._check_model_validity(section_model)
+
+    def validate(self):
+        """"
+        Validates the CDES, sections and registry forms
+        Raises an ExporterException in case of error
+        """
+        cdes = self._get_cdes(ExportType.REGISTRY_PLUS_CDES)
```
Comment (sztamas) [inline rdrf/rdrf/services/io/defs/exporter.py]: Not sure if this is good enough.

We should validate only the CDEs that will be included in the export, but there might be a mismatch if we don't pass in the export type to validate.

Also, I don't see where `form.complete_form_cdes` are validated.

Comment (alintreznai) [reply rdrf/rdrf/services/io/defs/exporter.py]: Updated. The complete_form_cdes are validated when the form is validated as it's a part of the form's clean method.

**Pull request merged**